In [ ]:
import numpy as np
import pandas as pd

from google.colab import files
uploaded = files.upload()

Saving temperature.csv to temperature (4).csv


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
import pandas as pd
file_path = '/content/temperature.csv'
df = pd.read_csv(file_path)
display(df.head())

record_id  month  day  year  AverageTemperatureFahr  \
0     474376      1    1  1853                     NaN   
1     474377      2    1  1853                     NaN   
2     474378      3    1  1853                     NaN   
3     474379      4    1  1853                     NaN   
4     474380      5    1  1853                     NaN   

   AverageTemperatureUncertaintyFahr      City country_id      Country  \
0                                NaN  Auckland        NEW  New Zealand   
1                                NaN  Auckland        NEW  New Zealand   
2                                NaN  Auckland        NEW  New Zealand   
3                                NaN  Auckland        NEW  New Zealand   
4                                NaN  Auckland        NEW  New Zealand   

  Latitude Longitude  
0   36.17S   175.03E  
1   36.17S   175.03E  
2   36.17S   175.03E  
3   36.17S   175.03E  
4   36.17S   175.03E

In [ ]:
df.describe()

record_id         month      day          year  \
count  4.847000e+04  48470.000000  48470.0  48470.000000   
mean   5.044463e+06      6.497050      1.0   1895.186590   
std    2.452494e+06      3.451431      0.0     73.646328   
min    4.743760e+05      1.000000      1.0   1743.000000   
25%    3.778360e+06      3.000000      1.0   1842.000000   
50%    4.672458e+06      6.000000      1.0   1901.000000   
75%    7.660919e+06      9.000000      1.0   1957.000000   
max    8.258730e+06     12.000000      1.0   2013.000000   

       AverageTemperatureFahr  AverageTemperatureUncertaintyFahr  
count            47541.000000                       47541.000000  
mean                51.755134                          34.334373  
std                 15.064100                           2.796225  
min                  2.856200                          32.088200  
25%                 40.172000                          32.621000  
50%                 54.593600                          33.161000  
75%                 63.708800                          34.973600  
max                 84.655400                          56.679800

In [ ]:
df2 = pd.DataFrame(df)
df2 = df2[((df2['month'] == 3) | (df2['month'] == 6) | (df2['month'] == 9) | (df2['month'] == 12))]
df2

record_id  month  day  year  AverageTemperatureFahr  \
2         474378      3    1  1853                     NaN   
5         474381      6    1  1853                 51.9062   
8         474384      9    1  1853                 52.5776   
11        474387     12    1  1853                 59.8460   
14        474390      3    1  1854                 64.9688   
...          ...    ...  ...   ...                     ...   
48457    8258718      9    1  2012                 57.1946   
48460    8258721     12    1  2012                 29.5070   
48463    8258724      3    1  2013                 29.9480   
48466    8258727      6    1  2013                 61.1672   
48469    8258730      9    1  2013                     NaN   

       AverageTemperatureUncertaintyFahr      City country_id      Country  \
2                                    NaN  Auckland        NEW  New Zealand   
5                                36.9572  Auckland        NEW  New Zealand   
8                                33.6380  Auckland        NEW  New Zealand   
11                               37.5062  Auckland        NEW  New Zealand   
14                               35.3966  Auckland        NEW  New Zealand   
...                                  ...       ...        ...          ...   
48457                            32.5094   Wroclaw        POL       Poland   
48460                            32.3132   Wroclaw        POL       Poland   
48463                            32.4536   Wroclaw        POL       Poland   
48466                            32.3330   Wroclaw        POL       Poland   
48469                                NaN   Wroclaw        POL       Poland   

      Latitude Longitude  
2       36.17S   175.03E  
5       36.17S   175.03E  
8       36.17S   175.03E  
11      36.17S   175.03E  
14      36.17S   175.03E  
...        ...       ...  
48457   50.63N    16.48E  
48460   50.63N    16.48E  
48463   50.63N    16.48E  
48466   50.63N    16.48E  
48469   50.63N    16.48E  

[16160 rows x 11 columns]

In [ ]:
import numpy as np

type(df2)

pandas.core.frame.DataFrame

In [ ]:
import numpy as np

np_array_from_df2 = df2.to_numpy()

average_temp_per_month = df2.groupby("month")["AverageTemperatureFahr"].mean()
print("Average Temperature par Month:")
average_temp_per_month

Average Temperature par Month:


month
3     44.312332
6     61.101684
9     59.459745
12    41.662069
Name: AverageTemperatureFahr, dtype: float64

# Task
Pivot the `df2` DataFrame to prepare for correlation analysis, by setting 'year' as the index, 'month' as columns, and 'AverageTemperatureFahr' as values. Then, calculate and display the correlation matrix of 'AverageTemperatureFahr' for months 3, 6, 9, and 12, and summarize the correlation analysis.

## Prepare Data for Correlation

### Subtask:
Pivot the 'df2' DataFrame to have 'year' as index, 'month' as columns, and 'AverageTemperatureFahr' as values, to prepare for correlation calculation between months.


**Reasoning**:
To prepare the data for correlation calculation between months, I need to pivot the 'df2' DataFrame to have 'year' as the index, 'month' as columns, and 'AverageTemperatureFahr' as the values, as specified in the instructions.



In [9]:
df_pivot = df2.pivot_table(index='year', columns='month', values='AverageTemperatureFahr')
display(df_pivot.head())

month        3         6         9         12
year                                         
1744        NaN  60.46556  57.32834  31.71902
1745   33.11888       NaN       NaN       NaN
1750   40.16318  62.11562  57.12926  29.91218
1751   39.12602  62.49740  56.18354       NaN
1752   38.63354       NaN       NaN  34.90970

## Calculate Correlation Matrix

### Subtask:
Calculate the correlation matrix between the 'AverageTemperatureFahr' for the months 3, 6, 9, and 12 from the prepared data.


**Reasoning**:
To calculate the correlation matrix, I will use the `.corr()` method on the `df_pivot` DataFrame, which already has 'year' as the index and the relevant months as columns with 'AverageTemperatureFahr' values.



In [10]:
correlation_matrix = df_pivot.corr()
display(correlation_matrix)

month        3         6         9         12
month                                        
3      1.000000 -0.428785  0.543921  0.838706
6     -0.428785  1.000000 -0.185564 -0.524021
9      0.543921 -0.185564  1.000000  0.554868
12     0.838706 -0.524021  0.554868  1.000000

## Final Task

### Subtask:
Summarize the correlation analysis between the average temperatures of the selected months.


## Summary:

### Q&A
The correlation analysis between the average temperatures of months 3 (March), 6 (June), 9 (September), and 12 (December) reveals distinct seasonal relationships. March and December exhibit a strong positive correlation (0.838706), while March and September, and September and December show moderate positive correlations (0.543921 and 0.554868, respectively). Conversely, June consistently displays negative correlations with March (-0.428785), September (-0.185564), and December (-0.524021), indicating an inverse relationship in average temperatures with these months.

### Data Analysis Key Findings
*   The `df2` DataFrame was successfully pivoted, setting 'year' as the index, 'month' as columns, and 'AverageTemperatureFahr' as values, preparing the data for correlation analysis.
*   A strong positive correlation (0.838706) was found between the average temperatures of March (month 3) and December (month 12).
*   Moderate positive correlations were observed between March (month 3) and September (month 9) at 0.543921, and between September (month 9) and December (month 12) at 0.554868.
*   June (month 6) exhibited negative correlations with all other selected months: -0.428785 with March (month 3), -0.185564 with September (month 9), and -0.524021 with December (month 12).

### Insights or Next Steps
*   The strong positive correlation between March and December, coupled with June's negative correlations with these months, suggests a clear seasonal temperature cycle where early and late year temperatures move in sync, opposite to mid-year temperatures.
*   Further analysis could involve exploring correlations for all months to gain a comprehensive understanding of yearly temperature patterns, or investigating if these correlations hold true across different geographical regions or climates.
